XML_to_SON/grammar_parser_test.py

```diff
@@ -186,60 +274,231 @@ def custom_serialize(obj, key_name="simulation", indent_size=0):
+    # To a string ("in_streams": "streams_", in the metadata). Solution has not been found before this PR.  
+    while isinstance(value, str):
+        reference_key = value
+        if reference_key in annotations:
+            value = annotations[reference_key]
+            print(f"Resolved '{reference_key}' to: {value} (type: {type(value)})")
+        else:
+            print(f"Warning: Reference key '{reference_key}' not found in annotations.")
+            return None
+
+    if not isinstance(value, dict):
+        print(f"Warning: Expected dictionary after resolving references, but found {type(value).__name__}.")
+        return None
+    return value
+
 def generate_doc_lines_for_key(key, value, annotations, annotation_key, doc_indent, child_indent):
+    # Still in progress to work with nested structures
     lines = []
-    optional = "(optional)" if value.get("MinOccurs", "1") == "0" else ""
-    type_str = f"[{value.get('ValType', 'Unknown')}]"
-    var_doc = annotations[annotation_key].get('vars', {}).get(key, {}).get('doc', 'No documentation available')
-    doc_lines = [f'%{optional} {type_str} {line}' for line in var_doc.split('\n')]
+    
+    if isinstance(value, dict):
+        value = resolve_reference(value, annotations)
+    else:
+        lines.append(f"{child_indent}{key} =\n")
+        return lines
+
+    annotation_namespace = annotations.get(annotation_key, {})
+    var_info = annotation_namespace.get('vars', {}).get(key, {})
+    if not isinstance(var_info, dict):
+        var_info = {}
+
+    optional = " (optional)" if value.get("MinOccurs", 1) == 0 else ""
+    val_type = value.get('ValType', var_info.get('type', 'Unknown'))
+    type_str = f"[{val_type}]"
+    var_doc = var_info.get('doc', 'No documentation available')
 
+    doc_lines = [f'%{optional} {type_str} {line}' for line in var_doc.split('\n')]
     lines.extend([f"{doc_indent}{line}" for line in doc_lines])
 
-    default_val = annotations[annotation_key].get('vars', {}).get(key, {}).get('default', '')
-    default_str = " = " + (str(default_val) if default_val != '' else "")
-    lines.append(f"{child_indent}{key}{default_str}\n")
+    default_val = var_info.get('default', None)
+    default_str = f" = {default_val}" if default_val is not None else " ="
+
+    if 'val' in value:
+        lines.append(f"{child_indent}{key} {{val{default_str}}}\n")
+    else:
+        lines.append(f"{child_indent}{key}{default_str}\n")
+
     return lines
 
 def custom_serialize_for_template(obj, annotations, key_name):
     lines = []
     indent_size = 0
     tab = 4 * " "
     base_indent = indent_size * " "
-    doc_indent = base_indent + 2 * tab 
+    doc_indent = base_indent + 2 * tab
     child_indent = base_indent + tab
 
-    matching_keys = [key for key in annotations.keys() if key.endswith(key_name)]
+    matching_keys = [key for key in annotations.keys() if key.split(":")[-2] == key_name.split("_")[-3] and key.split(":")[-1] == key_name.split("_")[-1]]
```

Comment:
I think this will be more readable if we define the following with variables first:
* `key_name.split("_")[-3]`
* `key_name.split("_")[-1]`